### PR TITLE
Fix double-printed logs after container restart

### LINKED ISSUE #217: Logs appear twice after container restart

### PR DESCRIPTION
**Pull request by lekkas**
Connects to #217 

Ref: https://github.com/resin-io-modules/resin-sync/issues/14

---
This change is [Reviewable](https://reviewable.io/reviews/resin-io/resin-supervisor/218)

### CODE REVIEW

**Comment by pcarranzav** [comment]
Good catch @lekkas - however:
* This will break logs when the container is restarted by docker. The logger.attach there is needed for that. Instead what should be done is somehow mark that `appHasDied[app.containerId]=false` when restarting via the supervisor endpoints.
* Please use "connects to  #issue" instead of "Closes #issue" in the PR opening comment for proper Waffle tracking.

**Comment by lekkas** [comment]
Ah right, I totally missed that. I'll look at it more closely because the underlying issue here appears to be a synchronisation one - both `docker.getEventsAsync()`and `application.start()` enter `attach()` and wait for the inner docker promises so neither manages to lock the other out.

Fixed the 'closes issue' as well.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218#-:-KNDCVbsfVT1WUqvooZZ:-730951431)*

**Comment by pcarranzav** [comment]
A better solution would be to ensure inside `logger.coffee` that we only attach once to each container.

**Comment by pcarranzav** [comment]
But actually... we currently do [here](https://github.com/resin-io/resin-supervisor/blob/master/src/lib/logger.coffee#L57)

**Comment by lekkas** [comment]
Well, this lock doesn't work as expected - because both tasks will pass this check and then wait on the promises (e.g. `docker.getContainer(app.containerId)`) . Then both tasks will enter our critical section at the same time, which will result in the duplicate logs.

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218#-:-KNDLej_QJ33PUeCF2sE:-1707431794)*

**Comment by pcarranzav** [comment]
Right, so we need to add a rwlock there

**Comment by lekkas** [comment]
Added

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218#-:-KNDWdxPyz8zwUU-IJCd:-502253829)*

File: src/lib/logger.coffee [hunk 1]
```diff
@@ -52,17 +53,25 @@ exports.log = ->
 	publish(arguments...)
 
 do ->
+	_lock = new Lock()
```
Comment (emirotin) [inline src/lib/logger.coffee]: I've recently seen the same logic in a different PR (though now it seems to be gone) — worth moving to utils module?

Comment (pcarranzav) [inline src/lib/logger.coffee]: Yeah, we have a bunch of rwlocks in several parts of the supervisor.
The one in logger was for a different purpose and is no longer necessary. I think it's fine to have this here.

**Comment by pcarranzav** [comment]
Please add a changelog entry. Other than that, LGTM :+1:

**Comment by Page-** [comment]
Reviewed 1 of 1 files at r1, 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218)*

**Comment by lekkas** [comment]
retest this please

**Comment by lekkas** [comment]
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/lib/logger.coffee, line 56 \[r2\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218#-KNDawPUMp_SUP3Ikmbl-r2-56:-KNGTJtRCDjkLMZtQRVW:676605422) ([raw file](https://github.com/resin-io/resin-supervisor/blob/efd3ed884950bcd065f9b1b6fd7937b46bd836a0/src/lib/logger.coffee#L56)):*
<details><summary><i>Previously, pcarranzav (Pablo Carranza Vélez) wrote…</i></summary>
> Yeah, we have a bunch of rwlocks in several parts of the supervisor.
> 
> The one in logger was for a different purpose and is no longer necessary. I think it's fine to have this here.
</details>
Cool

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218)*

**Comment by lekkas** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/lib/logger.coffee, line 56 \[r2\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218#-KNDawPUMp_SUP3Ikmbl-r2-56:-KNGTsLo2M6RbGi8uOQJ:-499146009) ([raw file](https://github.com/resin-io/resin-supervisor/blob/efd3ed884950bcd065f9b1b6fd7937b46bd836a0/src/lib/logger.coffee#L56)):*
<details><summary><i>Previously, lekkas (Kostas Lekkas) wrote…</i></summary>
> Cool
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/resin-io/resin-supervisor/218)*

**Pull request merged**